### ISSUE TITLE
"empty OP stack" for values running `dlv test`

### ISSUE THREAD

**Issue by postcert** [go-delve/delve#3640]
I originally saw this issue in `nvim` with `dap` but was able to reproduce it with the cli.
Running `dlv test` seems to be building the optimized/inlined version as the output in the cli is missing values.

```
❯ dlv test ./
Type 'help' for list of commands.
(dlv) break ./emulate_cycle.go:194
Breakpoint 1 set at 0x63cb9d for github.com/postcert/go8.(*Chip8).addRegistersWithCarry() ./emulate_cycle.go:194
(dlv) continue
> github.com/postcert/go8.(*Chip8).addRegistersWithCarry() ./emulate_cycle.go:194 (hits goroutine(36):1 total:1) (PC: 0x63cb9d)
   189:         chip.V[registerX>>8] ^= chip.V[registerY>>4]
   190: }
   191:
   192: func (chip *Chip8) addRegistersWithCarry(opcode uint16, value uint16) {
   193:         // TODO: 255 + 1 = 0, not 256 becuase of type. Failing overflow test below
=> 194:         temp := uint16(chip.V[opcode>>8]) + uint16(chip.V[opcode>>4])
   195:         if temp > 255 {
   196:                 chip.V[0xF] = 1
   197:         } else {
   198:                 chip.V[0xF] = 0
   199:         }
(dlv) stack -full
0  0x000000000063cb9d in github.com/postcert/go8.(*Chip8).addRegistersWithCarry
   at ./emulate_cycle.go:194
       chip = (unreadable empty OP stack)
       opcode = (unreadable empty OP stack)
       value = (unreadable empty OP stack)
...
```

but when I specify the `-N` or `-l` flags for the gcflags, it is able to pull the values.

```
❯ dlv test ./ --build-flags='-gcflags=all=-N'
Type 'help' for list of commands.
(dlv) break ./emulate_cycle.go:194
Breakpoint 1 set at 0x6ad940,0x6b2865 for (multiple functions)() ./emulate_cycle.go:194
(dlv) continue
> github.com/postcert/go8.TestAddRegistersWithCarry_Success() ./emulate_cycle.go:194 (hits goroutine(50):1 total:1) (PC: 0x6b2865)
Warning: debugging optimized function
   189:         chip.V[registerX>>8] ^= chip.V[registerY>>4]
   190: }
   191:
   192: func (chip *Chip8) addRegistersWithCarry(opcode uint16, value uint16) {
   193:         // TODO: 255 + 1 = 0, not 256 becuase of type. Failing overflow test below
=> 194:         temp := uint16(chip.V[opcode>>8]) + uint16(chip.V[opcode>>4])
   195:         if temp > 255 {
   196:                 chip.V[0xF] = 1
   197:         } else {
   198:                 chip.V[0xF] = 0
   199:         }
(dlv) stack -full
0  0x00000000006b2865 in github.com/postcert/go8.(*Chip8).addRegistersWithCarry
   at ./emulate_cycle.go:194
       chip = ("*github.com/postcert/go8.Chip8")(0xc000142640)
       opcode = 16
       value = 16

1  0x00000000006b2865 in github.com/postcert/go8.TestAddRegistersWithCarry_Success
   at ./emulate_cycle_test.go:254
       t = (*testing.T)(0xc0001c0820)
       chip = ("*github.com/postcert/go8.Chip8")(0xc000142640)
       opcode = 16
       value = 16
       chip = ("*github.com/postcert/go8.Chip8")(0xc000142640)
       temp = 108
...
(dlv) exit
```

1. What version of Delve are you using (`dlv version`)?
```
❯ dlv version && echo $(which dlv)
Delve Debugger
Version: 1.22.0
Build: $Id: 61ecdbbe1b574f0dd7d7bad8b6a5d564cce981e9 $
/usr/bin/dlv
```

2. What version of Go are you using? (`go version`)?
```
❯ go version && echo $(which go)
go version go1.21.6 linux/amd64
/usr/bin/go
```

3. What operating system and processor architecture are you using?
Arch Linux - x86_64

4. What did you do?
Requested `stack -full` to see the state variables outside of the current frame.

5. What did you expect to see?
The variable information as show when dlv is ran when I manually specify the build flags.

6. What did you see instead?
`unreadable empty OP stack` for all variables out of frame.

**Comment by postcert** [comment]
I'm at a bit of a loss on why specifying the `gcflags` parameter again is "fixing" it but at this point in the day I'm not going to make any sense of it.

**Comment by aarzilli** [comment]
Maybe because you are using a different value for `-gcflags`, delve uses `-gcflags='all=-N -l'` you are passing `-gcflags='all=-N'` thus enabling inlining. However I don't think this should make a difference in this case. Do you have a way for me to reproduce this?

**Comment by postcert** [comment]
Yeah, I’ll see if I can break it down to a minimal example and repro it on another machine/docker as well.

It’s just odd to me that the bare ‘delv test’ fails as well assuming it’s setting the desired flags.

**Comment by aarzilli** [comment]
Yes, it is odd. Especially given that it is not printing the optimization warning after `continue`.

**Comment by postcert** [comment]
I've created a repro case here: https://github.com/postcert/go8/tree/dlv_debug

Was able to reproduce the issue in Arch and Ubuntu containers with go1.21.6 and delve 1.22.0.

**Comment by aarzilli** [comment]
cc @dr2chase I think this is a compiler bug. This exact problem doesn't reproduce on go1.22 but similar instances of variables getting an empty location still do, they can be found by doing `objdump --dwarf exe | grep -B4 '0 byte', any executable compiled with `-gcflags=all=-N -l` will do. For example I can see it for the variable `runExitHook` in `runtime.runExitHooks`.

**Comment by aarzilli** [comment]
I've bisected this and it looks like this was (accidentally?) fixed by 505e50b1e34cdf6dff29615a076e26fb0780d10d https://go.dev/cl/509856. It only happens rarely and I could only find it in return parameters.

**Comment by tcortega** [comment]
> Closing this in favor of golang/go bug (also it's almost completely fixed in 1.22).

What was resolved after all? I'm facing the same issue. I am not using `dlv test` nor `dlv debug`. I'm using `dlv exec` after building the executable without passing any special flags.

**Comment by abitrolly** [comment]
@tcortega could be an outdated binary. I also got "empty OP stack" trying to `p err` when running `dlv exec` for compiled test. But then recompiled the test and the error had gone.

``````sh
$ go test -c -v ./commands/ci/status
$ ls -la status.test
-rwxr-xr-x. 1 anatoli anatoli 24420597 Mar  5 10:15 status.test
$ dlv exec status.test
...
(dlv)
> gitlab.com/gitlab-org/cli/commands/ci/status.NewCmdStatus.func1() ./commands/ci/status/status.go:67 (PC: 0xac3d40)
Warning: debugging optimized function
    62:                                 }
    63:                         } else {
    64:                                 branchConfig := git.ReadBranchConfig(branch)
    65:                                 if branchConfig.RemoteName == "" {
    66:                                         repo, err = f.BaseRepo()
=>  67:                                         if err != nil {
    68:                                                 return err
    69:                                         }
    70:                                         if branch == "" {
    71:                                                 branch, err = git.CurrentBranch()
    72:                                                 if err != nil {
(dlv) p err
error(*gitlab.com/gitlab-org/cli/internal/run.CmdError) *{
        Stderr: *bytes.Buffer {
                buf: []uint8 len: 139, cap: 1024, [102,97,116,97,108,58,32,110,111,116,32,97,32,103,105,116,32,114,101,112,111,115,105,116,111,114,121,32,40,111,114,32,97,110,121,32,112,97,114,101,110,116,32,117,112,32,116,111,32,109,111,117,110,116,32,112,111,105,110,116,32,47,41,10,...+75 more],
                off: 0,
                lastRead: opInvalid (0),},
        Args: []string len: 3, cap: 3, ["git","remote","-v"],
        Err: error(*os/exec.ExitError) *{
                ProcessState: *(*os.ProcessState)(0xc00054e0a8),
                Stderr: []uint8 len: 0, cap: 0, nil,},}
(dlv)
```
